torcharrow/dtypes_core.py

```diff
@@ -0,0 +1,342 @@
+
+
+# for now: no float16, and all date and time stuff, no categorical, (and Null is called Void)
+
+
+@dataclass(frozen=True)  # type: ignore
+class Numeric(DType):
+    pass
+
+
+@dataclass(frozen=True)
+class Boolean(DType):
+    nullable: bool = False
+    typecode: ty.ClassVar[str] = "b"
+    arraycode: ty.ClassVar[str] = "b"
+    name: ty.ClassVar[str] = "boolean"
+    default: ty.ClassVar[bool] = False
+
+    def constructor(self, nullable):
+        return Boolean(nullable)
+
+
+@dataclass(frozen=True)
+class Int8(Numeric):
+    nullable: bool = False
+    typecode: ty.ClassVar[str] = "c"
+    arraycode: ty.ClassVar[str] = "b"
+    name: ty.ClassVar[str] = "int8"
+    default: ty.ClassVar[int] = 0
+
+    def constructor(self, nullable):
+        return Int8(nullable)
+
+
+@dataclass(frozen=True)
+class Int16(Numeric):
+    nullable: bool = False
+    typecode: ty.ClassVar[str] = "s"
+    arraycode: ty.ClassVar[str] = "h"
+    name: ty.ClassVar[str] = "int16"
+    default: ty.ClassVar[int] = 0
+
+    def constructor(self, nullable):
+        return Int16(nullable)
+
+
+@dataclass(frozen=True)
+class Int32(Numeric):
+    nullable: bool = False
+    typecode: ty.ClassVar[str] = "i"
+    arraycode: ty.ClassVar[str] = "i"
+    name: ty.ClassVar[str] = "int32"
+    default: ty.ClassVar[int] = 0
+
+    def constructor(self, nullable):
+        return Int32(nullable)
+
+
+@dataclass(frozen=True)
+class Int64(Numeric):
+    nullable: bool = False
+    typecode: ty.ClassVar[str] = "l"
+    arraycode: ty.ClassVar[str] = "l"
+    name: ty.ClassVar[str] = "int64"
+    default: ty.ClassVar[int] = 0
+
+    def constructor(self, nullable):
+        return Int64(nullable)
+
+
+# Not all Arrow types are supported. We don't have a backend to support unsigned
+# integer types right now so they are removed to not confuse users. Feel free to
+# add unsigned int types when we have a supporting backend.
+
+
+@dataclass(frozen=True)
+class Float32(Numeric):
+    nullable: bool = False
+    typecode: ty.ClassVar[str] = "f"
+    arraycode: ty.ClassVar[str] = "f"
+    name: ty.ClassVar[str] = "float32"
+    default: ty.ClassVar[float] = 0.0
+
+    def constructor(self, nullable):
+        return Float32(nullable)
+
+
+@dataclass(frozen=True)
+class Float64(Numeric):
+    nullable: bool = False
+    typecode: ty.ClassVar[str] = "g"
+    arraycode: ty.ClassVar[str] = "d"
+    name: ty.ClassVar[str] = "float64"
+    default: ty.ClassVar[float] = 0.0
+
+    def constructor(self, nullable):
+        return Float64(nullable)
+
+
+@dataclass(frozen=True)
+class String(DType):
+    nullable: bool = False
+    typecode: ty.ClassVar[str] = "u"  # utf8 string (n byte)
+    arraycode: ty.ClassVar[str] = "w"  # wchar_t (2 byte)
+    name: ty.ClassVar[str] = "string"
+    default: ty.ClassVar[str] = ""
+
+    def constructor(self, nullable):
+        return String(nullable)
+
+
+@dataclass(frozen=True)
+class Map(DType):
+    key_dtype: DType
+    item_dtype: DType
```

Comment:
I think this `key/item` is coming from PyArrow's Map Type: https://arrow.apache.org/docs/python/generated/pyarrow.map_.html

But in many columnar Database (Presto/Velox) calls it `key/value` instead. `key/value` also seems to align with Pythonic API (in Python Dict, `item` means the (key, value) pair). 


Velox: https://github.com/facebookincubator/velox/blob/2148956b208132e4e37684392d28651e5ef9ff07/velox/vector/ComplexVector.h#L401-L402